src/main/java/com/google/sps/Scheduler.java

```diff
@@ -79,26 +79,50 @@ private void generateSchedulesHelper(List<Course> availableCourses, List<Schedul
-        .map(c -> c.getSection())
-        .collect(Collectors.toList());
+    List<Section> courseLectureSections = course.getLectureSections();
+    List<Section> courseLabSections = course.getLabSections();
+    boolean hasLabs = !courseLabSections.isEmpty();
+
+    List<Section> scheduleSections = new ArrayList<>();
+    for (ScheduledCourse c : currentSchedule) {
+      scheduleSections.add(c.getLectureSection());
+      if (c.getLabSection() != null) {
+        scheduleSections.add(c.getLabSection());
+      }
+    }
 
     /*
      * Iterate over each section in the current course and check for overlap. If the
      * section doesn't overlap and adding it would result in a valid schedule, add
      * it and make a new recursive call.
      */
-    for (Section section : courseSections) {
-      boolean doesNotOverlapSchedule = !sectionsOverlap(section, scheduleSections);
-      boolean doesNotExceedCreditLimit = 
-          currCredits + course.getCredits() <= invariants.getMaxCredits();
-      
-      if (doesNotOverlapSchedule && doesNotExceedCreditLimit) {
-        List<ScheduledCourse> newScheduledCourseLiist = new ArrayList<>(currentSchedule);
-        newScheduledCourseLiist.add(new ScheduledCourse(course, section));
-        generateSchedulesHelper(new ArrayList<>(availableCourses), 
-            newScheduledCourseLiist, invariants, generatedScheduledCourseLists);
+    if (hasLabs) {
+      for (Section lectureSection : courseLectureSections) {
+        for (Section labSection : courseLabSections) {
+          boolean doesNotOverlapSchedule = !sectionsOverlap(lectureSection, labSection, scheduleSections);
+          boolean doesNotExceedCreditLimit = 
+              currCredits + course.getCredits() <= invariants.getMaxCredits();
+          
+          if (doesNotOverlapSchedule && doesNotExceedCreditLimit) {
+            List<ScheduledCourse> newScheduledCourseList = new ArrayList<>(currentSchedule);
+            newScheduledCourseList.add(new ScheduledCourse(course, lectureSection, labSection));
+            generateSchedulesHelper(new ArrayList<>(availableCourses), 
+                newScheduledCourseList, invariants, generatedScheduledCourseLists);
+          }
+        }
+      }
+    } else {
+      for (Section lectureSection : courseLectureSections) {
+        boolean doesNotOverlapSchedule = !sectionsOverlap(lectureSection, scheduleSections);
+        boolean doesNotExceedCreditLimit = 
+            currCredits + course.getCredits() <= invariants.getMaxCredits();
+        
+        if (doesNotOverlapSchedule && doesNotExceedCreditLimit) {
+          List<ScheduledCourse> newScheduledCourseList = new ArrayList<>(currentSchedule);
+          newScheduledCourseList.add(new ScheduledCourse(course, lectureSection));
+          generateSchedulesHelper(new ArrayList<>(availableCourses), 
+              newScheduledCourseList, invariants, generatedScheduledCourseLists);
+        }
```

Comment:
This code is super redundant, but because the differences are pretty slight (and mostly in how calls are made), it feels more legible than other options?
Like theoretically I could move `doesNotExceedCreditLimit` outside of the larger if statement entirely, (as I write this I might do that as if that boolean is false looping over all the sections is completely unnecessary), but all of the other logic, despite being almost the same, is different enough that I don't know how I could abstract it further.

Reply:
Maybe you can keep the outer loop, `for(Section lectureSection : courseLectureSection)` as is (meaning do not replicate) and put the `if(hasLabs)...else...` inside that outer loop, so that you don't need to replicate the outerloop. Im not completely certain it could work, i.e. scope of variables, but something worth considering

Reply:
The indentation might be a bit off, but similar to what Achilles proposed:

```
for (Section lectureSection : courseLectureSections) {
  int totalCredits = currCredits + course.getCredits();
  boolean exceedsCreditLimit = totalCredits > invariants.getMaxCredits();

  if (exceedsCreditLimit) {
    continue;
  }

  for (Section labSection : courseLabSections) {
    boolean overlapSchedule = sectionsOverlap(
        lectureSection,
        labSection,
        scheduleSections);

    if (!overlapsSchedule) {
      ScheduledCourse = new ScheduledCourse(course, lectureSection, labSection);

      List<ScheduledCourse> newScheduledCourseList = new ArrayList<>(currentSchedule);
      newScheduledCourseList.add(scheduledCourse);
      
      generateSchedulesHelper(
          new ArrayList<>(availableCourses), 
          newScheduledCourseList,
          invariants,
          generatedScheduledCourseLists);
    }
  }

  if (!hasLabSection) { 
    boolean overlapSchedule = sectionsOverlap(lectureSection, scheduleSections);

    if (!overlapSchedule) {
      List<ScheduledCourse> newScheduledCourseList = new ArrayList<>(currentSchedule);
      newScheduledCourseList.add(new ScheduledCourse(course, lectureSection));

      generateSchedulesHelper(
          new ArrayList<>(availableCourses), 
          newScheduledCourseList,
          invariants,
          generatedScheduledCourseLists);
    }
  }
}
```

You could clean this up further by abstracting out the overlap check and recursive call logic.

Reply:
I didn't abstract out the overlap check/recursive call logic because the recursive call would require having the abstracted function take in all of the parameters for `generateSchedulesHelper` in addition to the parameters needed for the overlap check. Which would technically result in "cleaner" code, but a call to a helper-helper function with ~7 parameters felt uglier than just having 3-4 lines of redundant code.